alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+use std::cmp;
+
+use glutin::Rect;
+
+use alacritty_terminal::selection::SelectionRange;
+use alacritty_terminal::term::{DamagedLine, RenderableCursor, SizeInfo, TermDamage};
+use alacritty_terminal::vi_mode::ViModeCursor;
+
+/// Minimum part of damaged square when intersecting damaged rectangles.
+pub const MIN_DAMAGED_SQUARE_PART: f32 = 0.4;
+
+/// An error when we for look ahead in iterator, used to try to collect rectangles which pretty
+/// close to each other, but don't fit into `MIN_DAMAGED_SQUARE_PART` just by looking on the first
+/// next damaged rectangle.
+pub const MIN_DAMAGED_SQUARE_PART_ERR: f32 = MIN_DAMAGED_SQUARE_PART / 2.;
+
+pub struct RendererDamage<'a> {
+    index: usize,
```

Comment:
I'm not sure it's a good idea to have iterator indices stored on the struct itself? If so it definitely needs to be renamed to make things more clear. But I really don't like having mutable iterator state here.

Reply:
It's because we're doing look ahead. I can't just do `next()` all the time on terminal_damage, so I need to access it by index, advance ahead and fallback to previous values.

Reply:
If it's an iterator, it should be named appropriately. If it's just a data bag for all the damage information, that's different. It shouldn't be both. I feel like it's probably the former though, so a rename might be appropriate?